crates/sui-graphql-rpc/src/server/watermark_task.rs

```diff
@@ -155,36 +164,75 @@ impl Watermark {
         }
     }
 
+    #[allow(clippy::type_complexity)]
     pub(crate) async fn query(db: &Db) -> Result<Option<Watermark>, Error> {
         use checkpoints::dsl;
-        let Some((checkpoint, checkpoint_timestamp_ms, epoch)): Option<(i64, i64, i64)> = db
+        let Some(result): Option<Vec<(i64, i64, i64, Option<i64>)>> = db
             .execute(move |conn| {
                 async {
-                    conn.first(move || {
-                        dsl::checkpoints
-                            .select((dsl::sequence_number, dsl::timestamp_ms, dsl::epoch))
+                    conn.results(move || {
+                        let min_cp = dsl::checkpoints
+                            .select((
+                                dsl::sequence_number,
+                                dsl::timestamp_ms,
+                                dsl::epoch,
+                                dsl::min_tx_sequence_number,
+                            ))
+                            .order_by(dsl::sequence_number.asc())
+                            .limit(1);
+
+                        let max_cp = dsl::checkpoints
+                            .select((
+                                dsl::sequence_number,
+                                dsl::timestamp_ms,
+                                dsl::epoch,
+                                dsl::min_tx_sequence_number,
+                            ))
                             .order_by(dsl::sequence_number.desc())
+                            .limit(1);
+
+                        // Order by sequence_number, which is in the 1st position
+                        min_cp.union_all(max_cp).positional_order_by(1)
                     })
                     .await
                     .optional()
                 }
                 .scope_boxed()
             })
             .await
-            .map_err(|e| Error::Internal(format!("Failed to fetch checkpoint: {e}")))?
+            .map_err(|e| Error::Internal(format!("Failed to fetch watermark data: {e}")))?
         else {
+            // An empty response from the db is valid when indexer has not written any checkpoints
+            // to the db yet.
             return Ok(None);
         };
+
+        if result.len() != 2 {
+            return Err(Error::Internal(
+                "Expected 2 rows from watermark query".to_string(),
+            ));
+        }
+
+        let Some(lo_tx) = result[0].3 else {
+            return Err(Error::Internal(
+                "Expected min_tx_sequence_number to be non-null".to_string(),
+            ));
+        };
+
         Ok(Some(Watermark {
-            checkpoint: checkpoint as u64,
-            checkpoint_timestamp_ms: checkpoint_timestamp_ms as u64,
-            epoch: epoch as u64,
+            hi_cp: result[1].0 as u64,
+            hi_cp_timestamp_ms: result[1].1 as u64,
+            epoch: result[1].2 as u64,
+            lo_cp: result[0].0 as u64,
+            lo_tx: lo_tx as u64,
         }))
     }
```

Comment:
Better to structure it like this (fewer operations that can panic, handles the case where there is only one result without it being an error):

```rust
let (lo_cp, lo_tx) = if let Some((cp, _, _, Some(tx))) = result.first() {
  (cp, tx)
} else {
  return Err(Error::Internal("..."));
};

let (hi_cp, hi_cp_timestamp_ms, epoch) = if ((cp, timestamp_ms, epoch, _)) = result.last() {
  (cp, timestamp, epoch)
} else {
  return Err(Error::Internal("..."));
};

Ok(Some(Watermark {
    hi_cp,
    hi_cp_timestamp_ms,
    epoch,
    lo_cp,
    lo_tx,
}))
```